mlrun/runtimes/nuclio/api_gateway.py

```diff
@@ -0,0 +1,302 @@
+# Copyright 2023 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import base64
+from typing import Optional, Union
+from urllib.parse import urljoin
+
+import requests
+
+import mlrun
+import mlrun.common.schemas
+
+from .function import RemoteRuntime
+from .serving import ServingRuntime
+
+NUCLIO_API_GATEWAY_AUTHENTICATION_MODE_BASIC_AUTH = "basicAuth"
+NUCLIO_API_GATEWAY_AUTHENTICATION_MODE_NONE = "none"
+PROJECT_NAME_LABEL = "nuclio.io/project-name"
+
+
+class APIGateway:
+    def __init__(
+        self,
+        project,
+        name: str,
+        functions: Union[
+            list[str],
+            Union[
+                list[
+                    Union[
+                        RemoteRuntime,
+                        ServingRuntime,
+                    ]
+                ],
+                Union[RemoteRuntime, ServingRuntime],
+            ],
+        ],
+        description: str = "",
+        path: str = "/",
+        authentication_mode: Optional[
+            str
+        ] = NUCLIO_API_GATEWAY_AUTHENTICATION_MODE_NONE,
+        host: Optional[str] = None,
+        canary: Optional[list[int]] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
```

Comment:
instead of passing the auth mode + its options
read:
None -> ()
Basic -> (username, password)
IGZ -> (accessToken)
SSO -> (JWT)
and so on, just accept an argument that abstracts the authentication (an interface) mode. each concrete (none, basic, igz, sso) will implement some functionality that extends a given api gateway structure

Reply:
@liranbg this will be done in the following PR if you don't mind,  just not to add more classes/functionality to this one. It's already 1k changes. I've added link to this comment in jira task - https://jira.iguazeng.com/browse/ML-5372